src/sync/mod.rs

```diff
@@ -0,0 +1,355 @@
+            block_height
+        );
+
+        let mut signer = p2wsh::InputSigner::new(redeem_script);
+        let sign_input_messages = inputs
+            .iter()
+            .enumerate()
+            .map(|(index, proposal_input)| {
+                let signature = signer.sign_input(
+                    TxInRef::new(proposal.as_ref(), index),
+                    inputs[index].as_ref(),
+                    &(keypair.1).0.key,
+                )?;
+
+                signer.verify_input(
+                    TxInRef::new(proposal.as_ref(), index),
+                    proposal_input.as_ref(),
+                    &(keypair.0).0,
+                    &signature,
+                )?;
+
+                Ok(SignInput {
+                    transaction: proposal.clone(),
+                    input: index as u32,
+                    input_signature: signature.into(),
+                })
+            })
+            .collect::<Result<Vec<_>, failure::Error>>()
+            .map_err(AnchoringChainUpdateError::Internal)?;
+        // Send sign input transactions to the Exonum node.
+        for sign_input in sign_input_messages {
+            self.api_client
+                .sign_input(sign_input)
+                .wait()
+                .map_err(AnchoringChainUpdateError::Client)?;
+        }
+        Ok(())
+    }
+
+    fn find_private_key(
+        &self,
+        anchoring_keys: impl IntoIterator<Item = btc::PublicKey>,
+    ) -> Option<(btc::PublicKey, btc::PrivateKey)> {
+        anchoring_keys.into_iter().find_map(|public_key| {
+            self.key_pool
+                .get(&public_key)
+                .cloned()
+                .map(|private_key| (public_key, private_key))
+        })
+    }
+}
+
+/// Errors that occur when updating the sync with Bitcoin task.
+#[derive(Debug)]
+pub enum SyncWithBitcoinError<C: Display, R: Display> {
+    /// Error occurred in the private API client.
+    Client(C),
+    /// Error occurred in the Bitcoin relay.
+    Relay(R),
+    /// Internal error.
+    Internal(failure::Error),
+    /// Initial funding transaction is unconfirmed.
+    UnconfirmedFundingTransaction(Hash),
+}
+
+/// Pushes anchoring transactions to the Bitcoin blockchain.
+#[derive(Debug)]
+pub struct SyncWithBitcoinTask<T, R>
+where
+    T: PrivateApi + 'static,
+    R: BitcoinRelay + 'static,
+{
+    btc_relay: R,
+    api_client: T,
+}
+
+impl<T, R> SyncWithBitcoinTask<T, R>
+where
+    T: PrivateApi + 'static,
+    R: BitcoinRelay + 'static,
+    T::Error: Display,
+    R::Error: Display,
+{
+    /// Create a new sync with Bitcoin task instance.
+    pub fn new(btc_relay: R, api_client: T) -> Self {
+        Self {
+            api_client,
+            btc_relay,
+        }
+    }
+
+    /// Perform a one attempt to send the first uncommitted anchoring transaction into the Bitcoin network, if any.
+    /// sign an anchoring proposal, if any. Return an index of the first committed transaction.
+    pub fn process(
```

Comment:
I don't quite get the logic here. Wouldn't it be possible to send all uncommitted transactions to the relay at once? Why enforce a delay between sending transactions to the relay?

Reply:
Because bitcoind does not allow to push long chains of unconfirmed transactions. Thus it is more reasonable to push them by one by one.